neural_compressor/adaptor/torch_utils/teq.py

```diff
@@ -0,0 +1,391 @@
+                layer.weight = torch.nn.Parameter(
+                    weight, requires_grad=False)
+                bias = torch.zeros(layer.num_features, device=self.device, dtype=self.dtype)
+                layer.bias = torch.nn.Parameter(bias, requires_grad=False
+                                                )
+        elif isinstance(layer, torch.nn.LayerNorm):
+            if layer.elementwise_affine:
+                layer.weight *= scale
+                layer.bias *= scale
+            else:
+                layer.elementwise_affine = True
+                weight = torch.ones(layer.num_features, device=self.device, dtype=self.dtype) * scale
+                layer.weight = torch.nn.Parameter(
+                    torch.ones(weight, requires_grad=False))
+                bias = torch.zeros(layer.num_features, device=self.device, dtype=self.dtype)
+                layer.bias = torch.nn.Parameter(
+                    bias, requires_grad=False)
+
+        elif isinstance(layer, torch.nn.Conv2d):
+            ## the order could not be changed
+            if hasattr(layer, "bias") and (layer.bias != None):
+                layer.bias *= scale
+            scale = scale.view(scale.shape[0], 1, 1, 1)
+            layer.weight *= scale
+
+        elif isinstance(layer, torch.nn.Linear):
+            if hasattr(layer, "bias") and (layer.bias != None):
+                layer.bias *= scale
+            scale = scale.view(scale.shape[0], 1)
+            layer.weight *= scale
+
+
+        elif layer.__class__.__name__ == "LlamaRMSNorm" \
+                or layer.__class__.__name__ == "T5LayerNorm":  ##quite tricky
+            layer.weight *= scale
+
+        else:
+            logger.info(f"found unsupported layer {type(layer)}, try to multiply scale to "
+                f"weight and bias directly, this may introduce accuracy issue, please have a check ")
+            if hasattr(layer, "weight") and layer.weight != None:
+                layer.weight *= scale
+            if hasattr(layer, "bias") and layer.bias != None:
+                layer.bias *= scale
+
+    @torch.no_grad()
+    def _scale_layer_weight(self, layer, scale):  ##input channel
+        """
+        Scale the layer weights at input channel, depthwise conv output channel
+        :param layer_name: The layer name
+        :param scale: The scale to be multiplied
+        :return:
+        """
+        if layer.__class__.__name__ == "TEQMulLinear":
+            layer = layer.sq_linear
+
+        if layer.__class__.__name__ == "TEQLinearFakeQuant":
+            layer = layer.orig_layer
+
+        scale = scale.view(1, scale.shape[0])
+        layer.weight = torch.nn.Parameter(layer.weight * scale)
+        return scale
+
+    @torch.no_grad()
+    def transform(self):
+        """
+        apply alpha/scale
+        """
+        for ln_name, layer_names in self.absorb_to_layer.items():
```

Comment:
support customized absorb_to_layer later in case there are some bugs in auto-detection.

Reply:
ok